### PR TITLE
[PR] Adding basic table and metrics page

### LINKED ISSUE #245: Feat: Metrics!
At present I don't have _any_ metrics/stats for my usage of the `MVP`. 😞 

# Story

## `People` _Using_ the `MVP`

As a **`person`** _using_ the @dwyl `MVP` to test the features and give feedback https://github.com/dwyl/mvp/issues/140
I **`want`** a **quick way** of viewing basic aggregate stats on how many `items` and `timers` I have created
**`So` that** I know how much I'm using the `MVP` to improve my personal effectiveness.

## `People` _Building_ the `MVP`

As a **`person`** _building_ the @dwyl `MVP` to solve a _personal_ problem faced with existing task trackers or ["personal productivity" tools...] 
I **`want`** to know if the `MVP` is being used by _anyone_ 
**`So` that** I know if the features are being used.

# Todo

+ [ ] Create a new endpoint `/stats` 
+ [ ] Create a `LiveView` page that renders the data and is updated in real-time
+ [ ] The `stats` channel should only be used for _aggregate_ data i.e. nothing personally identifiable.

Build a basic metrics dashboard that displays:

+ [ ] `person_id` 
+ [ ] Number of `items` 
+ [ ] Number of `timers` 

+ [ ] Display the data as a `<table`> (we can create pretty graphs later!)

> **Note**: this will be an _internal_ dashboard for now and will remain "anonymous".
But in the future it will be a **_personal_ dashboard** for the `person` _using_ the `App` to see how much and _when_ they are using it. 

So we will want a _visual_ representation of their usage similar to what `GitHub` has with their contribution graph:

image

image

We will encourage people to have "streaks" or at least to know they are making progress toward their goals.
This is "Step 1" on the Roadmap: [github.com/dwyl/product-roadmap#why-are-we-building-an-app](https://github.com/dwyl/product-roadmap/tree/02e07b0ab15cbe2337ccabb9de686496dba6a6fa#why-are-we-building-an-app)
Knowing that you are making **_daily_ progress** towards your goals is the **secret** to personal effectiveness and staying focussed long-term and not getting distracted ... See: [mattcutts.com/blog/seinfeld-calendar-trick](https://www.mattcutts.com/blog/seinfeld-calendar-trick/)

Longer term this will be a feature in itself. But for now we just need something basic. 🙏

### PR DESCRIPTION
**Pull request by LuchoTurtle**
closes #245 

Adds a simple table with the `person_id`, number of `items` and numbers of `timers` created:

### CODE REVIEW

**Comment by nelsonic** [comment]
Review App: https://mvp-pr-250.fly.dev not fully loading (doesn't connect to web sockets...)
image

See: https://fly.io/apps/mvp-pr-250/monitoring

**Comment by nelsonic** [comment]
@SimonLab could you please take a quick look at this and understand why the coverage is missing that line? 🔍 
And if possible get the review app working too. 🤞

**Comment by SimonLab** [comment]
Codecov seems broken for me:
image

Same for you?

**Comment by LuchoTurtle** [comment]
Can't access either :/ 

image

Which is weird, Codecov says [it's fully operational](https://status.codecov.com/).

Had to check Twitter or something to see if we're alone but a handful of people seem to have the same problem. 

image

**Comment by SimonLab** [comment]
Re run the tests, coverage updated

**Comment by SimonLab** [comment]
Related to the review apps, I think the socket connection issue is linked to how fly deployed the application (my latest PR review apps has been deployed and work, which make me think the issue is specific to this application and hopefully just a one of).
You can still access the `stats` endpoint https://mvp-pr-250.fly.dev/stats

However because review apps create a new database at the moment the table is empty
image

File: lib/app/item.ex [hunk 1]
```diff
@@ -168,6 +168,32 @@ defmodule App.Item do
...
+
+  iex> person_with_item_and_timer_count()
+  [
+    %{name: nil, num_items: 3, num_timers: 8, person_id: 0}
+    %{name: username, num_items: 1, num_timers: 3, person_id: 1}
+  ]
+  """
+  def person_with_item_and_timer_count() do
+    sql = """
+    select p.person_id, p.name, COUNT(distinct i.id) as "num_items", count(distinct t.id) as "num_timers"
```
Comment (nelsonic) [inline lib/app/item.ex:185]: If we're writing raw  `SQL`, prefer the keywords e.g. `SELECT`, `AS`, `FROM` to be `ALLCAPS` for readability.

e.g: 
https://github.com/dwyl/mvp/blob/faba54760ab0435653434aec7c9f801096fbbd8b/lib/app/item.ex#L151-L154

File: lib/app_web/live/stats_live.html.heex [hunk 2]
```diff
@@ -0,0 +1,46 @@
+<main class="font-sans container mx-auto">
+  <div class="relative overflow-x-auto mt-12">
+    <h1 class="mb-12 text-xl font-extrabold leading-none tracking-tight text-gray-900 md:text-5xl lg:text-6xl dark:text-white">
+      User metrics
```
Comment (nelsonic) [inline lib/app_web/live/stats_live.html.heex:4]: We are very deliberately avoiding the word "User" in our projects; https://github.com/dwyl/app/issues/33

File: lib/app/item.ex [hunk 3]
```diff
@@ -168,6 +168,32 @@ defmodule App.Item do
...
+    from people p
```
Comment (nelsonic) [inline lib/app/item.ex:186]: The `people` table should no longer exist in this project. https://github.com/dwyl/mvp/issues/118
Removing it _now_. ✂️

**Comment by nelsonic** [comment]
This PR introduces a bunch of new warnings: https://github.com/dwyl/mvp/actions/runs/3876133208/jobs/6624083418#step:8:27
image

This creates unnecessary noise in the terminal. 🙉 

Cleaning it up now. 🧹

File: lib/app/item.ex (hunk 1 above)
Comment (nelsonic) [reply lib/app/item.ex]: Simplified: https://github.com/dwyl/mvp/issues/245#issuecomment-1377238534

File: lib/app_web/live/stats_live.html.heex (hunk 2 above)
Comment (nelsonic) [reply lib/app_web/live/stats_live.html.heex]: removed. ✂️

**Comment by nelsonic** [comment]
Sadly the review app is useless for this PR because no data is being stored in the DB so the stats table is empty:
image

**Comment by nelsonic** [comment]
https://app.codecov.io/gh/dwyl/mvp/pull/250
image

✅

**Review by nelsonic — APPROVED** [review]
Looks great! Thanks @LuchoTurtle 🐢 🚀

**Comment by nelsonic** [comment]
Need to look at why https://mvp.fly.dev/stats is `404` tomorrow morning ...

image

🤷‍♂️

**Pull request merged**